Source/VisualStudioTools/Private/VSTestAdapterCommandlet.cpp

```diff
@@ -0,0 +1,177 @@
+// Copyright 2022 (c) Microsoft. All rights reserved.
+
+#include "VSTestAdapterCommandlet.h"
+
+#include <Runtime/Core/Public/Async/TaskGraphInterfaces.h>
+#include <Runtime/Core/Public/Containers/Ticker.h>
+#include <fstream>
+#include <string>
+
+#include "VisualStudioTools.h"
+
+int32 UVSTestAdapterCommandlet::Main(const FString& Params)
+{
+	TArray<FString> Tokens;
+	TArray<FString> Switches;
+	TMap<FString, FString> ParamVals;
+
+	// Functionality for Unreal Engine Test Adapter.
+	ParseCommandLine(*Params, Tokens, Switches, ParamVals);
+	FAutomationTestFramework::GetInstance().SetRequestedTestFilter( EAutomationTestFlags::PriorityMask | EAutomationTestFlags::ProductFilter | EAutomationTestFlags::SmokeFilter | EAutomationTestFlags::PerfFilter | EAutomationTestFlags::StressFilter | EAutomationTestFlags::NegativeFilter );	
+	if ( ParamVals.Contains( TEXT( "listtests" ) ) )
+	{
+		return ListTests( ParamVals[TEXT( "listtests" )] );
+	}
+	else if ( ParamVals.Contains( TEXT( "runtests" ) ) && ParamVals.Contains( TEXT( "testresultfile" ) ) )
+	{
+		return RunTests( ParamVals[TEXT( "runtests" )], ParamVals[TEXT( "testresultfile" )] );
+	}
+
+	return 0;
+}
+
+
+int32 UVSTestAdapterCommandlet::ListTests( const FString& TargetFile )
+{
+	std::wofstream OutFile( *TargetFile );
+	if ( !OutFile.good() )
+	{
+		UE_LOG( LogVisualStudioTools, Error, TEXT( "Failed to open file at path: %s" ), *TargetFile );
+		return 1;
+	}
+
+	FAutomationTestFramework& Framework = FAutomationTestFramework::GetInstance();
+
+	TArray< FAutomationTestInfo > TestInfos;
+	GetAllTests( TestInfos );
+
+	for ( const auto& TestInfo : TestInfos )
+	{
+		const FString TestCommand = TestInfo.GetTestName();
+		const FString DisplayName = TestInfo.GetDisplayName();
+		const FString SourceFile = TestInfo.GetSourceFile();
+		const int32 Line = TestInfo.GetSourceFileLine();
+
+		OutFile << *TestCommand << TEXT( "|" ) << *DisplayName << TEXT( "|" ) << Line << TEXT( "|" ) << *SourceFile << std::endl;
```

Comment:
This (and below for `RunTests`) should ensure to escape the individual fields from verbatim "|" used in the test names. (Handling and unescaping then needs to happen on the reader side too.)

Reply:
Will address in future PR soon.